### ISSUE TITLE
DrawImage Bitmap AntiAliases layers

### ISSUE THREAD

**Issue by jaebird** [picoe/Eto#679]
This is a bit hard to describe, but here I go:

If I draw all elements with the Drawable's main Graphics context the resulting elements are displayed correctly. If I create a transparent bitmap and then draw elements onto the bitmap, when the bitmap is drawn onto the drawable, the bitmap elements appear antialiased or dimmer. It seems like pixel edges have an alpha component to them even though I drew them with AntiAlias off using colors with the alpha value set to max (1.0 or 0xFF).

My desire is that the results of both methods are the exact same output. Any clues of where to look? Is this a cairo issue? Or floating point rounding?

If I draw a white line without antialiasing, I would expect it to be the same thickness and color from a transparent backed bitmap or drawing directly the Drawable with the Graphics context.

Using 2.3-build-0856 on Ubuntu with Gtk2, but Gtk3 does the same thing.

**Comment by jaebird** [comment]
Could it be scaling the image by a subpixels?

**Comment by cra0zy** [comment]
I think I know what you are talking about, never tried drawing images with Eto drawing API, but with Gtk it does use cairo and cairo has several modes in which it will draw stuff: https://www.cairographics.org/manual/cairo-cairo-pattern-t.html#cairo-filter-t

The needed filter for getting image to be drawn the way they are is `CAIRO_FILTER_FAST` but that's not the default filter and I'm guessing Eto is not setting it.

The code for Gtk# 3 would look something like:
```c#
_surface = new ImageSurface(filePath);
_pattern = new SurfacePattern(_surface);
_pattern.Filter = Filter.Fast;

private void DrawingArea_Drawn(object o, DrawnArgs args)
{
    var context = args.Cr;

    context.SetSource(_pattern);
    context.Rectangle(...);
    context.Fill();
    context.Stroke();
}
```

**Comment by cra0zy** [comment]
Looking at the source code, Eto does have a way of setting this, `e.Graphics.ImageInterpolation = ImageInterpolation.Low;` should fix your issue.

**Comment by jaebird** [comment]
I was using ImageInterpolation.None, I guess that is different than Low. I'll look at the source.

**Comment by cra0zy** [comment]
Oh, now that I read the enums it definitely seems that they were rotated, #697

**Comment by cra0zy** [comment]
I can still be wrong on this I haven't use Cairo API in a long time....

**Comment by jaebird** [comment]
I tested with the current version using ImageInterpolation.Low (which is rotated). Seems like both None and Low work very much the same way and both exhibit the issue I described. Any other tricks up your sleeve? Right now the primary issue is the font rendering, which aliases the text against a transparent background. Then when layered, the black backround is mixed with the alpha of the font, causing it to "dim". This may just be a limitation of how I'm using the graphics. Chroma keying would be the better way to go.

**Comment by cra0zy** [comment]
Sorry, out of ideas.

**Comment by jaebird** [comment]
So should the enums be changed? The FAST is the fastest one and should be
referenced by ImageInterpolation.None I would think. I think your PR is
good: https://github.com/picoe/Eto/pull/697

CAIRO_FILTER_FAST: A high-performance filter, with quality similar to
CAIRO_FILTER_NEAREST
CAIRO_FILTER_GOOD: A reasonable-performance filter, with quality similar to
CAIRO_FILTER_BILINEAR
CAIRO_FILTER_BEST: The highest-quality available, performance may not be
suitable for interactive use.
CAIRO_FILTER_NEAREST: Nearest-neighbor filtering
CAIRO_FILTER_BILINEAR: Linear interpolation in two dimensions
CAIRO_FILTER_GAUSSIAN: This filter value is currently unimplemented, and
should not be used in current code.

I'm seeing similar performance with both, but I haven't tested in depth yet.

What do you think?

**Comment by cwensley** [comment]
@jaebird I don't think so.  Eto's ImageInterpolation.None should mean nearest neighbour, it doesn't mean it should be "fastest".

If we don't care about accuracy, we could add new Eto "Fast" enum value and map that to whatever is actually fastest on each platform.  However, it really depends on so many factors (e.g. what you're drawing, if you're doing scaling, etc), that it will always need per-app tweaking to determine what works best so it may be misleading.

**Comment by cra0zy** [comment]
Cairo API is very bad with the naming conventions, first half is named after FAST/GOOD/BEST, while the second half is named after actual filters... it really makes it hard to arrange them properly.

That being said CAIRO_FILTER_NEAREST is using nearest-neighbor filtering which should as the name implies just fill the image with pixel from neighbors so it being mapped to ImageInterpolation.None should be correct.

I actually have no idea what the FAST/GOOD/BEST since the docs don't actually describe them :/ Also last time I needed pixel perfect drawing with Cairo I used FAST filter, but I think I only used it because I didn't notice the NEAREST filter at that time.

**Comment by jaebird** [comment]
I understand the comments. What is weird is the FAST filter says the
quality is similar to NEAREST, so maybe they are the same in the cairo src.
I didn't see a noticeable difference with them for my use case.
